gateway-js/src/legacy/IntrospectAndCompose.ts

```diff
@@ -0,0 +1,148 @@
+import {
+  composeAndValidate,
+  compositionHasErrors,
+  ServiceDefinition,
+} from '@apollo/federation';
+import { Logger } from 'apollo-server-types';
+import CallableInstance from 'callable-instance';
+import { HeadersInit } from 'node-fetch';
+import {
+  ServiceEndpointDefinition,
+  SupergraphSdlHook,
+  SupergraphSdlUpdateFunction,
+} from '..';
+import {
+  getServiceDefinitionsFromRemoteEndpoint,
+  Service,
+} from './loadServicesFromRemoteEndpoint';
+import { waitUntil } from '../utilities/waitUntil';
+
+export interface IntrospectAndComposeOptions {
+  subgraphs: ServiceEndpointDefinition[];
+  introspectionHeaders?:
+    | HeadersInit
+    | ((
+        service: ServiceEndpointDefinition,
+      ) => Promise<HeadersInit> | HeadersInit);
+  pollIntervalInMs?: number;
+  logger?: Logger;
+}
+
+type State =
+  | { phase: 'initialized' }
+  | { phase: 'polling'; pollingPromise?: Promise<void> }
+  | { phase: 'stopped' };
+
+export class IntrospectAndCompose extends CallableInstance<
+  Parameters<SupergraphSdlHook>,
+  ReturnType<SupergraphSdlHook>
+> {
+  private config: IntrospectAndComposeOptions;
+  private update?: SupergraphSdlUpdateFunction;
+  private subgraphs?: Service[];
+  private serviceSdlCache: Map<string, string> = new Map();
+  private pollIntervalInMs?: number;
+  private timerRef: NodeJS.Timeout | null = null;
+  private state: State;
+
+  constructor(options: IntrospectAndComposeOptions) {
+    super('instanceCallableMethod');
+
+    this.config = options;
+    this.pollIntervalInMs = options.pollIntervalInMs;
+    this.state = { phase: 'initialized' };
+  }
+
+  // @ts-ignore noUsedLocals
+  private async instanceCallableMethod(
```

Comment:
This probably should be `protected`.

That said, as we discussed on zoom, perhaps it will be popular for folks to implement this hook as a class so maybe we should just have a 1-method interface instead. I also think it feels a little surprising in TS to pass `somethingDocumentedAsAFunction: new Class()` anyway...

Reply:
Implemented object awareness as part of the `supergraphSdl` API, so now if an object is passed in with an `initialize` method it will be called as if it was a `supergraphSdl` function.

c0b3c40